File: kotlin/src/test/kotlin/DataStructuresTest.kt
import datastructures.*
import kotlin.test.*

class DataStructuresTest {
    // Stack Tests
    @Test
    fun `stack operations`() {
        val stack = Stack<Int>()
        assertTrue(stack.isEmpty())

        stack.push(1)
        stack.push(2)
        stack.push(3)

        assertEquals(3, stack.size())
        assertEquals(3, stack.peek())
        assertEquals(3, stack.pop())
        assertEquals(2, stack.pop())
        assertEquals(1, stack.size())

        stack.clear()
        assertTrue(stack.isEmpty())
    }

    // Queue Tests
    @Test
    fun `queue operations`() {
        val queue = Queue<Int>()
        assertTrue(queue.isEmpty())

        queue.enqueue(1)
        queue.enqueue(2)
        queue.enqueue(3)

        assertEquals(3, queue.size())
        assertEquals(1, queue.peek())
        assertEquals(1, queue.dequeue())
        assertEquals(2, queue.dequeue())
        assertEquals(1, queue.size())
    }

    // DynamicArray Tests
    @Test
    fun `dynamic array operations`() {
        val arr = DynamicArray<Int>()
        assertTrue(arr.isEmpty())

        arr.push(10)
        arr.push(20)
        arr.push(30)

        assertEquals(3, arr.size())
        assertEquals(10, arr.get(0))
        assertEquals(30, arr.get(2))

        arr.set(1, 25)
        assertEquals(25, arr.get(1))

        assertEquals(30, arr.pop())
        assertEquals(2, arr.size())

        arr.insert(1, 15)
        assertEquals(listOf(10, 15, 25), arr.toList())

        arr.removeAt(1)
        assertEquals(listOf(10, 25), arr.toList())
    }

    // SinglyLinkedList Tests
    @Test
    fun `singly linked list operations`() {
        val list = SinglyLinkedList<Int>()
        assertTrue(list.isEmpty())

        list.append(1)
        list.append(2)
        list.prepend(0)

        assertEquals(3, list.size())
        assertEquals(0, list.get(0))
        assertEquals(2, list.get(2))

        list.insert(1, 10)
        assertEquals(listOf(0, 10, 1, 2), list.toList())

        list.removeAt(1)
        assertEquals(listOf(0, 1, 2), list.toList())

        assertTrue(list.contains(1))
        assertFalse(list.contains(100))
    }

    // DoublyLinkedList Tests
    @Test
    fun `doubly linked list operations`() {
        val list = DoublyLinkedList<Int>()
        assertTrue(list.isEmpty())

        list.append(1)
        list.append(2)
        list.prepend(0)

        assertEquals(3, list.size())
        assertEquals(0, list.getFirst())
        assertEquals(2, list.getLast())

        assertEquals(0, list.removeFirst())
        assertEquals(2, list.removeLast())
        assertEquals(1, list.size())
    }

    // Deque Tests
    @Test
    fun `deque operations`() {
        val deque = Deque<Int>()
        assertTrue(deque.isEmpty())

        deque.pushBack(2)
        deque.pushFront(1)
        deque.pushBack(3)

        assertEquals(3, deque.size())
        assertEquals(1, deque.peekFront())
        assertEquals(3, deque.peekBack())

        assertEquals(1, deque.popFront())
        assertEquals(3, deque.popBack())
        assertEquals(1, deque.size())
    }

    // HashTable Tests
    @Test
    fun `hash table operations`() {
        val table = HashTable<String, Int>()
        assertTrue(table.isEmpty())

        table.put("one", 1)
        table.put("two", 2)
        table.put("three", 3)

        assertEquals(3, table.size())
        assertEquals(1, table.get("one"))
        assertEquals(2, table.get("two"))
        assertNull(table.get("four"))

        assertTrue(table.contains("one"))
        assertFalse(table.contains("four"))

        table.put("one", 100)
        assertEquals(100, table.get("one"))

        assertEquals(2, table.remove("two"))
        assertEquals(2, table.size())
        assertFalse(table.contains("two"))
    }

    // BinarySearchTree Tests
    @Test
    fun `binary search tree operations`() {
        val bst = BinarySearchTree<Int>()
        assertTrue(bst.isEmpty())

        bst.insert(5)
        bst.insert(3)
        bst.insert(7)
        bst.insert(1)
        bst.insert(9)

        assertEquals(5, bst.size())
        assertTrue(bst.contains(5))
        assertTrue(bst.contains(3))
        assertFalse(bst.contains(100))

        assertEquals(1, bst.min())
        assertEquals(9, bst.max())

        assertEquals(listOf(1, 3, 5, 7, 9), bst.inOrder())

        assertTrue(bst.remove(3))
        assertFalse(bst.contains(3))
        assertEquals(listOf(1, 5, 7, 9), bst.inOrder())
    }

    // MinHeap Tests
    @Test
    fun `min heap operations`() {
        val heap = MinHeap<Int>()
        assertTrue(heap.isEmpty())

        heap.insert(5)
        heap.insert(3)
        heap.insert(7)
        heap.insert(1)
        heap.insert(9)

        assertEquals(5, heap.size())
        assertEquals(1, heap.peek())

        assertEquals(1, heap.extractMin())
        assertEquals(3, heap.extractMin())
        assertEquals(5, heap.extractMin())
        assertEquals(7, heap.extractMin())
        assertEquals(9, heap.extractMin())

        assertTrue(heap.isEmpty())
    }

    @Test
    fun `min heap with custom comparator`() {
        // Max heap using custom comparator
        val heap = MinHeap<Int>(compareByDescending { it })

        heap.insert(5)
        heap.insert(3)
        heap.insert(7)

        assertEquals(7, heap.extractMin())
        assertEquals(5, heap.extractMin())
        assertEquals(3, heap.extractMin())
    }

    // DisjointSet Tests
    @Test
    fun `disjoint set operations`() {
        val ds = DisjointSet(5)

        assertEquals(5, ds.count())
        assertFalse(ds.connected(0, 1))

        ds.union(0, 1)
        assertTrue(ds.connected(0, 1))
        assertEquals(4, ds.count())

        ds.union(2, 3)
        ds.union(0, 2)

        assertTrue(ds.connected(0, 3))
        assertTrue(ds.connected(1, 2))
        assertEquals(2, ds.count())

        assertFalse(ds.connected(0, 4))
    }
}

File: elixir/test/data_structures_test.exs
defmodule DataStructuresTest do
  use ExUnit.Case, async: true

  alias DataStructures.{
    Stack,
    Queue,
    DynamicArray,
    SinglyLinkedList,
    DoublyLinkedList,
    Deque,
    HashTable,
    BinarySearchTree,
    MinHeap,
    DisjointSet
  }

  # Stack Tests
  describe "Stack" do
    test "push and pop work correctly" do
      stack =
        Stack.new()
        |> Stack.push(1)
        |> Stack.push(2)
        |> Stack.push(3)

      assert {:ok, 3, stack} = Stack.pop(stack)
      assert {:ok, 2, stack} = Stack.pop(stack)
      assert {:ok, 1, _stack} = Stack.pop(stack)
    end

    test "peek returns top element without removing" do
      stack = Stack.new() |> Stack.push(42)
      assert {:ok, 42} = Stack.peek(stack)
      assert Stack.size(stack) == 1
    end

    test "empty? returns correct value" do
      assert Stack.empty?(Stack.new())
      refute Stack.empty?(Stack.new() |> Stack.push(1))
    end
  end

  # Queue Tests
  describe "Queue" do
    test "enqueue and dequeue work correctly" do
      queue =
        Queue.new()
        |> Queue.enqueue(1)
        |> Queue.enqueue(2)
        |> Queue.enqueue(3)

      assert {:ok, 1, queue} = Queue.dequeue(queue)
      assert {:ok, 2, queue} = Queue.dequeue(queue)
      assert {:ok, 3, _queue} = Queue.dequeue(queue)
    end

    test "maintains FIFO order" do
      queue =
        Queue.new()
        |> Queue.enqueue(1)
        |> Queue.enqueue(2)
        |> Queue.enqueue(3)

      assert Queue.to_list(queue) == [1, 2, 3]
    end
  end

  # DynamicArray Tests
  describe "DynamicArray" do
    test "push and pop work correctly" do
      arr =
        DynamicArray.new()
        |> DynamicArray.push(1)
        |> DynamicArray.push(2)
        |> DynamicArray.push(3)

      assert {:ok, 3, arr} = DynamicArray.pop(arr)
      assert DynamicArray.size(arr) == 2
    end

    test "get and set work correctly" do
      arr = DynamicArray.from_list([1, 2, 3])
      assert {:ok, 2} = DynamicArray.get(arr, 1)
      assert {:ok, updated} = DynamicArray.set(arr, 1, 42)
      assert {:ok, 42} = DynamicArray.get(updated, 1)
    end
  end

  # SinglyLinkedList Tests
  describe "SinglyLinkedList" do
    test "prepend and append work correctly" do
      list =
        SinglyLinkedList.new()
        |> SinglyLinkedList.prepend(2)
        |> SinglyLinkedList.prepend(1)
        |> SinglyLinkedList.append(3)

      assert SinglyLinkedList.to_list(list) == [1, 2, 3]
    end

    test "index_of finds elements" do
      list = SinglyLinkedList.from_list([10, 20, 30])
      assert SinglyLinkedList.index_of(list, 20) == 1
      assert SinglyLinkedList.index_of(list, 100) == nil
    end
  end

  # DoublyLinkedList Tests
  describe "DoublyLinkedList" do
    test "prepend and append work correctly" do
      list =
        DoublyLinkedList.new()
        |> DoublyLinkedList.prepend(2)
        |> DoublyLinkedList.prepend(1)
        |> DoublyLinkedList.append(3)

      assert DoublyLinkedList.to_list(list) == [1, 2, 3]
    end

    test "first and last work correctly" do
      list = DoublyLinkedList.from_list([1, 2, 3])
      assert {:ok, 1} = DoublyLinkedList.first(list)
      assert {:ok, 3} = DoublyLinkedList.last(list)
    end
  end

  # Deque Tests
  describe "Deque" do
    test "push_front and push_back work correctly" do
      deque =
        Deque.new()
        |> Deque.push_front(2)
        |> Deque.push_front(1)
        |> Deque.push_back(3)

      assert Deque.to_list(deque) == [1, 2, 3]
    end

    test "pop_front and pop_back work correctly" do
      deque = Deque.from_list([1, 2, 3])
      assert {:ok, 1, _} = Deque.pop_front(deque)
      assert {:ok, 3, _} = Deque.pop_back(deque)
    end
  end

  # HashTable Tests
  describe "HashTable" do
    test "put and get work correctly" do
      table =
        HashTable.new()
        |> HashTable.put("one", 1)
        |> HashTable.put("two", 2)

      assert {:ok, 1} = HashTable.get(table, "one")
      assert {:error, :not_found} = HashTable.get(table, "three")
    end

    test "remove works correctly" do
      table = HashTable.new() |> HashTable.put("key", 42)
      {value, updated} = HashTable.remove(table, "key")
      assert value == 42
      assert {:error, :not_found} = HashTable.get(updated, "key")
    end
  end

  # BinarySearchTree Tests
  describe "BinarySearchTree" do
    test "insert and contains? work correctly" do
      bst = BinarySearchTree.from_list([5, 3, 7, 1, 9])
      assert BinarySearchTree.contains?(bst, 5)
      assert BinarySearchTree.contains?(bst, 3)
      refute BinarySearchTree.contains?(bst, 10)
    end

    test "find_min and find_max work correctly" do
      bst = BinarySearchTree.from_list([5, 3, 7, 1, 9])
      assert {:ok, 1} = BinarySearchTree.find_min(bst)
      assert {:ok, 9} = BinarySearchTree.find_max(bst)
    end

    test "inorder traversal returns sorted list" do
      bst = BinarySearchTree.from_list([5, 3, 7, 1, 9])
      assert BinarySearchTree.inorder(bst) == [1, 3, 5, 7, 9]
    end
  end

  # MinHeap Tests
  describe "MinHeap" do
    test "insert and extract_min work correctly" do
      heap = MinHeap.from_list([5, 3, 7, 1, 9])
      assert {:ok, 1, heap} = MinHeap.extract_min(heap)
      assert {:ok, 3, _heap} = MinHeap.extract_min(heap)
    end

    test "to_list returns sorted order" do
      heap = MinHeap.from_list([5, 3, 7, 1, 9])
      assert MinHeap.to_list(heap) == [1, 3, 5, 7, 9]
    end

    test "peek returns minimum without removing" do
      heap = MinHeap.from_list([5, 3, 7])
      assert {:ok, 3} = MinHeap.peek(heap)
      assert MinHeap.size(heap) == 3
    end
  end

  # DisjointSet Tests
  describe "DisjointSet" do
    test "make_set and find work correctly" do
      ds =
        DisjointSet.new()
        |> DisjointSet.make_set(1)
        |> DisjointSet.make_set(2)

      assert {:ok, 1, _} = DisjointSet.find(ds, 1)
      assert {:error, :not_found} = DisjointSet.find(ds, 3)
    end

    test "union and connected? work correctly" do
      ds =
        DisjointSet.new()
        |> DisjointSet.make_set(1)
        |> DisjointSet.make_set(2)
        |> DisjointSet.make_set(3)

      refute DisjointSet.connected?(ds, 1, 2)

      {:ok, ds} = DisjointSet.union(ds, 1, 2)
      assert DisjointSet.connected?(ds, 1, 2)
      refute DisjointSet.connected?(ds, 1, 3)
    end

    test "set_count tracks disjoint sets" do
      ds =
        DisjointSet.new()
        |> DisjointSet.make_set(1)
        |> DisjointSet.make_set(2)
        |> DisjointSet.make_set(3)
        |> DisjointSet.make_set(4)

      assert DisjointSet.set_count(ds) == 4

      {:ok, ds} = DisjointSet.union(ds, 1, 2)
      assert DisjointSet.set_count(ds) == 3

      {:ok, ds} = DisjointSet.union(ds, 3, 4)
      assert DisjointSet.set_count(ds) == 2
    end
  end
end
